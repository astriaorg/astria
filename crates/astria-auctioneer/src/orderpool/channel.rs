use jiff::Timestamp;
use tokio::sync::mpsc;
use uuid::Uuid;

use super::{
    in_memory::{
        InsertedOrReplaced,
        RemovedOrNotFound,
    },
    Order,
};

pub(crate) struct Request {
    pub(super) order: Order,
    pub(super) to_requester: tokio::sync::oneshot::Sender<Response>,
}

pub(crate) struct ForOrder {
    pub(crate) uuid: Uuid,
    pub(crate) action: InsertedOrReplaced,
}
pub(crate) struct ForCancellation {
    pub(crate) uuid: Uuid,
    pub(crate) timestamp: Timestamp,
    pub(crate) action: RemovedOrNotFound,
}

pub(crate) enum Response {
    ForOrder(ForOrder),
    ForCancellation(ForCancellation),
}

impl From<ForCancellation> for Response {
    fn from(value: ForCancellation) -> Self {
        Self::ForCancellation(value)
    }
}
impl From<ForOrder> for Response {
    fn from(value: ForOrder) -> Self {
        Self::ForOrder(value)
    }
}
pub(super) fn new() -> (Sender, Receiver) {
    let (tx, rx) = mpsc::channel(32);
    (
        Sender {
            inner: tx,
        },
        Receiver {
            inner: rx,
        },
    )
}

pub(super) struct Receiver {
    inner: mpsc::Receiver<Request>,
}

impl Receiver {
    pub(super) async fn recv(&mut self) -> Option<Request> {
        self.inner.recv().await
    }
}

#[derive(Debug, thiserror::Error)]
pub(crate) enum SendError {
    #[error("no available capacity in order pool")]
    Full,
    #[error("order pool channel is closed")]
    Closed,
    #[error("order pool dropped the response channel before sending a response")]
    Dropped,
}

impl From<tokio::sync::mpsc::error::TrySendError<Request>> for SendError {
    fn from(value: tokio::sync::mpsc::error::TrySendError<Request>) -> Self {
        match value {
            mpsc::error::TrySendError::Full(_) => Self::Full,
            mpsc::error::TrySendError::Closed(_) => Self::Closed,
        }
    }
}

impl From<tokio::sync::oneshot::error::RecvError> for SendError {
    fn from(_value: tokio::sync::oneshot::error::RecvError) -> Self {
        Self::Dropped
    }
}

#[derive(Clone)]
pub(crate) struct Sender {
    inner: mpsc::Sender<Request>,
}

impl Sender {
    pub(crate) async fn send(&self, order: Order) -> Result<Response, SendError> {
        let (tx, rx) = tokio::sync::oneshot::channel();
        self.inner.try_send(Request {
            order,
            to_requester: tx,
        })?;
        let rsp = rx.await?;
        Ok(rsp)
    }
}
