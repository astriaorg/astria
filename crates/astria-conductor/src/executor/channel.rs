//! An mpsc channel bounded by an externally driven semaphore.
//!
//! While the main purpose of this channel is to send [`sequencer_client::SequencerBlock`]s
//! from a sequencer reader to the executor, the channel is generic over the values that are
//! being sent to better test its functionality.

use std::sync::{
    Arc,
    Weak,
};

use tokio::sync::{
    mpsc::{
        error::SendError as TokioSendError,
        unbounded_channel,
        UnboundedReceiver,
        UnboundedSender,
    },
    AcquireError,
    Semaphore,
    TryAcquireError,
};

/// Creates an mpsc channel for sending soft blocks between asynchronous task.
///
/// The initial bound of the channel is 0 and the receiver is expected to add
/// capacity to the channel.
pub(super) fn soft_block_channel<T>() -> (Sender<T>, Receiver<T>) {
    let cap = 0;
    let sem = Arc::new(Semaphore::new(0));
    let (tx, rx) = unbounded_channel();
    let sender = Sender {
        chan: tx,
        sem: Arc::downgrade(&sem),
    };
    let receiver = Receiver {
        cap,
        chan: rx,
        sem,
    };
    (sender, receiver)
}

#[derive(Debug, thiserror::Error, PartialEq)]
#[error("the channel is closed")]
pub(crate) struct SendError;

impl From<AcquireError> for SendError {
    fn from(_: AcquireError) -> Self {
        Self
    }
}

impl<T> From<TokioSendError<T>> for SendError {
    fn from(_: TokioSendError<T>) -> Self {
        Self
    }
}

// allow: this is mimicking tokio's `SendError` that returns the stack-allocated object.
#[allow(clippy::result_large_err)]
#[derive(Debug, thiserror::Error, PartialEq)]
pub(crate) enum TrySendError<T> {
    #[error("the channel is closed")]
    Closed(T),
    #[error("no permits available")]
    NoPermits(T),
}

impl<T> TrySendError<T> {
    fn from_semaphore(err: &TryAcquireError, block: T) -> Self {
        match err {
            tokio::sync::TryAcquireError::Closed => Self::Closed(block),
            tokio::sync::TryAcquireError::NoPermits => Self::NoPermits(block),
        }
    }
}

impl<T> From<TokioSendError<T>> for TrySendError<T> {
    fn from(err: TokioSendError<T>) -> Self {
        Self::Closed(err.0)
    }
}

#[derive(Debug, Clone)]
pub(super) struct Sender<T> {
    sem: Weak<Semaphore>,
    chan: UnboundedSender<T>,
}

impl<T> Sender<T> {
    /// Sends a block, waiting until the channel has permits.
    ///
    /// Returns an error if the channel is closed.
    pub(super) async fn send(&self, block: T) -> Result<(), SendError> {
        let sem = self.sem.upgrade().ok_or(SendError)?;
        let permit = sem.acquire().await?;
        permit.forget();
        self.chan.send(block)?;
        Ok(())
    }

    /// Attempts to send a block without blocking.
    ///
    /// Returns an error if the channel is out of permits or if it has been closed.
    // allow: this is mimicking tokio's `TrySendError` that returns the stack-allocated object.
    #[allow(clippy::result_large_err)]
    pub(super) fn try_send(&self, block: T) -> Result<(), TrySendError<T>> {
        let sem = match self.sem.upgrade() {
            None => return Err(TrySendError::Closed(block)),
            Some(sem) => sem,
        };
        let permit = match sem.try_acquire() {
            Err(err) => return Err(TrySendError::from_semaphore(&err, block)),
            Ok(permit) => permit,
        };
        permit.forget();
        self.chan.send(block)?;
        Ok(())
    }
}

pub(super) struct Receiver<T> {
    cap: usize,
    sem: Arc<Semaphore>,
    chan: UnboundedReceiver<T>,
}

impl<T> Drop for Receiver<T> {
    fn drop(&mut self) {
        self.sem.close();
    }
}

impl<T> Receiver<T> {
    /// Sets the channel's capacity to `cap`.
    ///
    /// `cap` will be the maximum number of blocks that can be sent
    /// over the channel before new permits are added with `[SoftBlockReceiver::add_permits]`.
    pub(super) fn set_capacity(&mut self, cap: usize) {
        self.cap = cap;
    }

    /// Adds up to `capacity` number of permits to the channel.
    ///
    /// `capacity` is previously set by [`SoftBlockReceiver::set_capacity`]
    /// or zero by default.
    pub(super) fn fill_permits(&self) {
        let additional = self.cap.saturating_sub(self.sem.available_permits());
        self.sem.add_permits(additional);
    }

    /// Receives a block over the channel.
    pub(super) async fn recv(&mut self) -> Option<T> {
        self.chan.recv().await
    }
}

#[cfg(test)]
mod tests {
    use super::{
        soft_block_channel,
        SendError,
        TrySendError,
    };

    #[test]
    fn fresh_channel_has_no_capacity() {
        let (tx, _rx) = soft_block_channel::<()>();
        assert_eq!(
            tx.try_send(()).unwrap_err(),
            TrySendError::NoPermits(()),
            "a fresh channel starts without permits"
        );
    }

    #[test]
    fn permits_are_filled_to_capacity() {
        let cap = 2;
        let (tx, mut rx) = soft_block_channel::<()>();
        rx.set_capacity(cap);
        rx.fill_permits();
        for _ in 0..cap {
            tx.try_send(()).expect("the channel should have capacity");
        }
        assert_eq!(
            tx.try_send(()).unwrap_err(),
            TrySendError::NoPermits(()),
            "a channel that has its permits used up should return with a NoPermits error until \
             refilled or closed",
        );
    }

    #[test]
    fn refilling_twice_has_no_effect() {
        let cap = 2;
        let (tx, mut rx) = soft_block_channel::<()>();
        rx.set_capacity(cap);
        rx.fill_permits();
        rx.fill_permits();
        for _ in 0..cap {
            tx.try_send(()).expect("the channel should have capacity");
        }
        assert_eq!(
            tx.try_send(()).unwrap_err(),
            TrySendError::NoPermits(()),
            "refilling twice in a row should result in the same number of permits"
        );
    }

    #[test]
    fn try_sending_to_dropped_receiver_returns_closed_error() {
        let (tx, rx) = soft_block_channel::<()>();
        std::mem::drop(rx);
        assert_eq!(
            tx.try_send(()).unwrap_err(),
            TrySendError::Closed(()),
            "a channel with a dropped receiver is considered closed",
        );
    }

    #[tokio::test]
    async fn async_sending_to_dropped_receiver_returns_closed_error() {
        let (tx, rx) = soft_block_channel::<()>();
        std::mem::drop(rx);
        assert_eq!(
            tx.send(()).await.unwrap_err(),
            SendError,
            "a channel with a dropped receiver is considered closed",
        );
    }

    #[tokio::test]
    #[should_panic(expected = "receiving with all senders dropped should return None")]
    async fn receiving_without_any_remaining_receivers_returns_none() {
        let (tx, mut rx) = soft_block_channel::<()>();
        std::mem::drop(tx);
        rx.recv()
            .await
            .expect("receiving with all senders dropped should return None");
    }
}
