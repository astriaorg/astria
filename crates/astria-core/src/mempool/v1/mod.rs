pub mod transaction_status;
