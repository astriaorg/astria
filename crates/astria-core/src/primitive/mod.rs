pub use pbjson_types::Timestamp;
pub mod v1;
