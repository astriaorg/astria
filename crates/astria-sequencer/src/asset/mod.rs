use std::sync::OnceLock;

use proto::native::sequencer::v1alpha1::asset::Denom;

pub(crate) static NATIVE_ASSET: OnceLock<Denom> = OnceLock::new();

pub(crate) fn initialize_native_asset(native_asset: &str) {
    if NATIVE_ASSET.get().is_some() {
        tracing::error!("native asset should only be set once");
        return;
    }

    let denom = Denom::from_base_denom(native_asset);
    NATIVE_ASSET
        .set(denom)
        .expect("native asset should only be set once");
}

pub(crate) fn get_native_asset() -> &'static Denom {
    NATIVE_ASSET.get().expect("native asset should be set")
}
