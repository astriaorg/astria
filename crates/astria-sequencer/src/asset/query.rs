use anyhow::Context as _;
use astria_core::{
    primitive::v1::asset,
    protocol::{
        abci::AbciErrorCode,
        asset::v1alpha1::AllowedFeeAssetIdsResponse,
    },
};
use cnidarium::Storage;
use prost::Message as _;
use tendermint::abci::{
    request,
    response,
};

use crate::{
    asset::state_ext::StateReadExt as _,
    state_ext::StateReadExt,
};

// Retrieve the full asset denomination given the asset ID.
//
// Example:
// `abci-cli query --path=asset/denom/<DENOM_ID>`
pub(crate) async fn denom_request(
    storage: Storage,
    request: request::Query,
    params: Vec<(String, String)>,
) -> response::Query {
    use astria_core::protocol::asset::v1alpha1::DenomResponse;

    // use the latest snapshot, as this is a lookup of id->denom
    let snapshot = storage.latest_snapshot();
    let asset_id = match preprocess_request(&params) {
        Ok(asset_id) => asset_id,
        Err(err_rsp) => return err_rsp,
    };

    let height = match snapshot.get_block_height().await {
        Ok(height) => height,
        Err(err) => {
            return response::Query {
                code: AbciErrorCode::INTERNAL_ERROR.into(),
                info: AbciErrorCode::INTERNAL_ERROR.to_string(),
                log: format!("failed getting block height: {err:#}"),
                ..response::Query::default()
            };
        }
    };

    let maybe_denom = match snapshot.get_ibc_asset(asset_id).await {
        Ok(maybe_denom) => maybe_denom,
        Err(err) => {
            return response::Query {
                code: AbciErrorCode::INTERNAL_ERROR.into(),
                info: AbciErrorCode::INTERNAL_ERROR.to_string(),
                log: format!("failed to retrieve denomination `{asset_id}`: {err:#}"),
                ..response::Query::default()
            };
        }
    };

    let Some(denom) = maybe_denom else {
        return response::Query {
            code: AbciErrorCode::VALUE_NOT_FOUND.into(),
            info: AbciErrorCode::VALUE_NOT_FOUND.to_string(),
            log: format!("failed to retrieve value for denomination ID`{asset_id}`"),
            ..response::Query::default()
        };
    };

    let payload = DenomResponse {
        height,
        denom,
    }
    .into_raw()
    .encode_to_vec()
    .into();

    let height = tendermint::block::Height::try_from(height).expect("height must fit into an i64");
    response::Query {
        code: tendermint::abci::Code::Ok,
        key: request.path.into_bytes().into(),
        value: payload,
        height,
        ..response::Query::default()
    }
}

fn preprocess_request(params: &[(String, String)]) -> anyhow::Result<asset::Id, response::Query> {
    let Some(asset_id) = params.iter().find_map(|(k, v)| (k == "id").then_some(v)) else {
        return Err(response::Query {
            code: AbciErrorCode::INVALID_PARAMETER.into(),
            info: AbciErrorCode::INVALID_PARAMETER.to_string(),
            log: "path did not contain asset ID parameter".into(),
            ..response::Query::default()
        });
    };
    let asset_id = hex::decode(asset_id)
        .context("failed decoding hex encoded bytes")
        .and_then(|addr| {
            asset::Id::try_from_slice(&addr).context("failed constructing asset ID from bytes")
        })
        .map_err(|err| response::Query {
            code: AbciErrorCode::INVALID_PARAMETER.into(),
            info: AbciErrorCode::INVALID_PARAMETER.to_string(),
            log: format!("asset ID could not be constructed from provided parameter: {err:#}"),
            ..response::Query::default()
        })?;
    Ok(asset_id)
}

pub(crate) async fn allowed_fee_asset_ids_request(
    storage: Storage,
    request: request::Query,
    _params: Vec<(String, String)>,
) -> response::Query {
    // get last snapshot
    let snapshot = storage.latest_snapshot();

    // get height from snapshot
    let height = match snapshot.get_block_height().await {
        Ok(height) => height,
        Err(err) => {
            return response::Query {
                code: AbciErrorCode::INTERNAL_ERROR.into(),
                info: AbciErrorCode::INTERNAL_ERROR.to_string(),
                log: format!("failed getting block height: {err:#}"),
                ..response::Query::default()
            };
        }
    };

    // get ids from snapshot at height
    let fee_asset_ids = match snapshot.get_allowed_fee_assets().await {
        Ok(maybe_fee_assets) => maybe_fee_assets,
        Err(err) => {
            return response::Query {
                code: AbciErrorCode::INTERNAL_ERROR.into(),
                info: AbciErrorCode::INTERNAL_ERROR.to_string(),
                log: format!("failed to retrieve allowed fee assets: {err:#}"),
                ..response::Query::default()
            };
        }
    };

    let payload = AllowedFeeAssetIdsResponse {
        height,
        fee_asset_ids,
    }
    .into_raw()
    .encode_to_vec()
    .into();

    let height = tendermint::block::Height::try_from(height).expect("height must fit into an i64");
    response::Query {
        code: tendermint::abci::Code::Ok,
        key: request.path.into_bytes().into(),
        value: payload,
        height,
        ..response::Query::default()
    }
}
