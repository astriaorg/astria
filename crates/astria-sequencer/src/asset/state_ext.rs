use anyhow::{
    bail,
    Context as _,
    Result,
};
use astria_core::primitive::v1::{
    asset,
    asset::Denom,
};
use async_trait::async_trait;
use borsh::{
    BorshDeserialize,
    BorshSerialize,
};
use cnidarium::{
    StateRead,
    StateWrite,
};
use hex::ToHex as _;
use tracing::instrument;

/// Newtype wrapper to read and write a denomination trace from rocksdb.
#[derive(BorshSerialize, BorshDeserialize, Debug)]
struct DenominationTrace(String);

fn asset_storage_key(asset: asset::Id) -> String {
    format!("asset/{}", asset.encode_hex::<String>())
}

#[async_trait]
pub(crate) trait StateReadExt: StateRead {
    #[instrument(skip(self))]
    async fn has_ibc_asset(&self, id: asset::Id) -> Result<bool> {
        match self
            .get_raw(&asset_storage_key(id))
            .await
            .context("failed reading raw asset from state")?
        {
            Some(_) => Ok(true),
            None => Ok(false),
        }
    }

    #[instrument(skip(self))]
    async fn get_ibc_asset(&self, id: asset::Id) -> Result<Denom> {
        let Some(bytes) = self
            .get_raw(&asset_storage_key(id))
            .await
            .context("failed reading raw asset from state")?
        else {
            bail!("asset not found");
        };

        let DenominationTrace(denom_str) =
            DenominationTrace::try_from_slice(&bytes).context("invalid asset bytes")?;
        let denom: Denom = denom_str.into();
        Ok(denom)
    }
}

impl<T: ?Sized + StateRead> StateReadExt for T {}

#[async_trait]
pub(crate) trait StateWriteExt: StateWrite {
    #[instrument(skip(self))]
    fn put_ibc_asset(&mut self, id: asset::Id, asset: &Denom) -> Result<()> {
        let bytes = borsh::to_vec(&DenominationTrace(asset.denomination_trace()))
            .context("failed to serialize asset")?;
        self.put_raw(asset_storage_key(id), bytes);
        Ok(())
    }
}

impl<T: StateWrite> StateWriteExt for T {}

#[cfg(test)]
mod test {
    use astria_core::primitive::v1::asset::{
        Denom,
        Id,
    };
    use cnidarium::StateDelta;

    use super::{
        StateReadExt as _,
        StateWriteExt as _,
    };

    #[tokio::test]
    async fn get_ibc_asset_non_existent() {
        let storage = cnidarium::TempStorage::new().await.unwrap();
        let snapshot = storage.latest_snapshot();
        let state = StateDelta::new(snapshot);

        let asset = Id::from_denom("asset");

        // gets for non existing assets fail
        state
            .get_ibc_asset(asset)
            .await
            .expect_err("gets for non existing ibc assets should fail");
    }

    #[tokio::test]
    async fn has_ibc_asset() {
        let storage = cnidarium::TempStorage::new().await.unwrap();
        let snapshot = storage.latest_snapshot();
        let mut state = StateDelta::new(snapshot);

        let denom = Denom::from_base_denom("asset");

        // non existing calls are ok for 'has'
        assert!(
            !state
                .has_ibc_asset(denom.id())
                .await
                .expect("'has' for non existing ibc assets should be ok"),
            "query for non existing asset should return false"
        );

        state
            .put_ibc_asset(denom.id(), &denom)
            .expect("putting ibc asset should not fail");

        // existing calls are ok for 'has'
        assert!(
            state
                .has_ibc_asset(denom.id())
                .await
                .expect("'has' for existing ibc assets should be ok"),
            "query for existing asset should return true"
        );
    }

    #[tokio::test]
    async fn put_ibc_asset_simple() {
        let storage = cnidarium::TempStorage::new().await.unwrap();
        let snapshot = storage.latest_snapshot();
        let mut state = StateDelta::new(snapshot);

        // can write new
        let denom = Denom::from_base_denom("asset");
        state
            .put_ibc_asset(denom.id(), &denom)
            .expect("putting ibc asset should not fail");
        assert_eq!(
            state
                .get_ibc_asset(denom.id())
                .await
                .expect("an ibc asset was written and must exist inside the database"),
            denom,
            "stored ibc asset was not what was expected"
        );
    }

    #[tokio::test]
    async fn put_ibc_asset_complex() {
        let storage = cnidarium::TempStorage::new().await.unwrap();
        let snapshot = storage.latest_snapshot();
        let mut state = StateDelta::new(snapshot);

        // can write new
        let denom = Denom::from_base_denom("asset_0");
        state
            .put_ibc_asset(denom.id(), &denom)
            .expect("putting ibc asset should not fail");
        assert_eq!(
            state
                .get_ibc_asset(denom.id())
                .await
                .expect("an ibc asset was written and must exist inside the database"),
            denom,
            "stored ibc asset was not what was expected"
        );

        // can write another without affecting original
        let denom_1 = Denom::from_base_denom("asset_1");
        state
            .put_ibc_asset(denom_1.id(), &denom_1)
            .expect("putting ibc asset should not fail");
        assert_eq!(
            state
                .get_ibc_asset(denom_1.id())
                .await
                .expect("an additional ibc asset was written and must exist inside the database"),
            denom_1,
            "additional ibc asset was not what was expected"
        );
        assert_eq!(
            state
                .get_ibc_asset(denom.id())
                .await
                .expect("an ibc asset was written and must exist inside the database"),
            denom,
            "original ibc asset was not what was expected"
        );
    }

    #[tokio::test]
    async fn put_ibc_asset_can_write_unrelated_ids_to_denoms() {
        let storage = cnidarium::TempStorage::new().await.unwrap();
        let snapshot = storage.latest_snapshot();
        let mut state = StateDelta::new(snapshot);

        // can write unrelated ids and denoms
        let id_key = Id::from_denom("asset_0");
        let denom = Denom::from_base_denom("asset_1");
        state
            .put_ibc_asset(id_key, &denom)
            .expect("putting ibc asset should not fail");

        // see that id key and denom's stored id differ
        assert_ne!(
            state
                .get_ibc_asset(id_key)
                .await
                .expect("an ibc asset was written and must exist inside the database")
                .id(),
            id_key,
            "stored ibc asset was not what was expected"
        );
    }
}
