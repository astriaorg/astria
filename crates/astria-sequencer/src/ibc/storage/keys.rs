use std::borrow::Cow;

use astria_core::primitive::v1::asset::IbcPrefixed;
use ibc_types::core::channel::ChannelId;

use crate::{
    accounts::AddressBytes,
    storage::keys::{
        AccountPrefixer,
        Asset,
    },
};

pub(in crate::ibc) const IBC_SUDO: &str = "ibc/sudo";
const IBC_RELAYER_PREFIX: &str = "ibc/relayer/";

/// Example: `ibc/channel-xxx/balance/ibc/0101....0101`.
///                      |int|           |64 hex chars|
pub(in crate::ibc) fn channel_balance<'a, TAsset>(channel: &ChannelId, asset: &'a TAsset) -> String
where
    &'a TAsset: Into<Cow<'a, IbcPrefixed>>,
{
    format!("ibc/{channel}/balance/{}", Asset::from(asset))
}

/// Example: `ibc/relayer/gGhH....zZ4=`.
///                      |base64 chars|
pub(in crate::ibc) fn ibc_relayer<T: AddressBytes>(address: &T) -> String {
    AccountPrefixer::new(IBC_RELAYER_PREFIX, address).to_string()
}

#[cfg(test)]
mod tests {
    use astria_core::primitive::v1::{
        asset::Denom,
        Address,
    };

    use super::*;

    const COMPONENT_PREFIX: &str = "ibc/";

    fn channel_id() -> ChannelId {
        ChannelId::new(5)
    }

    fn address() -> Address {
        "astria1rsxyjrcm255ds9euthjx6yc3vrjt9sxrm9cfgm"
            .parse()
            .unwrap()
    }

    fn asset() -> Denom {
        "an/asset/with/a/prefix".parse().unwrap()
    }

    #[test]
    fn keys_should_not_change() {
        insta::assert_snapshot!(IBC_SUDO);
        insta::assert_snapshot!(channel_balance(&channel_id(), &asset()));
        insta::assert_snapshot!(ibc_relayer(&address()));
    }

    #[test]
    fn keys_should_have_component_prefix() {
        assert!(IBC_SUDO.starts_with(COMPONENT_PREFIX));
        assert!(channel_balance(&channel_id(), &asset()).starts_with(COMPONENT_PREFIX));
        assert!(ibc_relayer(&address()).starts_with(COMPONENT_PREFIX));
    }
}
