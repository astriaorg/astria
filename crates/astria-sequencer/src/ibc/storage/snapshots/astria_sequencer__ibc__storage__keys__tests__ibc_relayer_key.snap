---
source: crates/astria-sequencer/src/ibc/storage/keys.rs
expression: ibc_relayer(&address())
---
ibc/relayer/HAxJDxtVKNgXPF3kbRMRYOSywMM=
